src/assets/assets.service.ts

```diff
@@ -0,0 +1,61 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { Injectable, NotFoundException } from '@nestjs/common';
+import { BasePrismaClient } from '../prisma/types/base-prisma-client';
+import { CreateAssetOptions } from './interfaces/create-asset-options';
+import { Asset, Transaction } from '.prisma/client';
+
+@Injectable()
+export class AssetsService {
+  async findByIdentifierOrThrow(
+    identifier: string,
+    prisma: BasePrismaClient,
+  ): Promise<Asset> {
+    const record = await prisma.asset.findUnique({
+      where: {
+        identifier,
+      },
+    });
+    if (!record) {
+      throw new NotFoundException();
+    }
+    return record;
+  }
+
+  async upsert(
+    options: CreateAssetOptions,
+    transaction: Transaction,
+    prisma: BasePrismaClient,
+  ): Promise<Asset> {
+    return prisma.asset.upsert({
+      create: {
+        identifier: options.identifier,
+        metadata: options.metadata,
+        name: options.name,
+        owner: options.owner,
+        supply: BigInt(0),
+        created_transaction_id: transaction.id,
+      },
+      update: {},
```

Comment:
I am not familiar with how the `update` vs `where` clause works with `prisma`. Can you update test to make sure that this doesn't update all assets (only one with provided `identifier`)?

Reply:
Prisma has a unique set of args based on the schema. For assets, it is:

```ts
  export type AssetWhereUniqueInput = {
    id?: number
    identifier?: string
  }
```

Since we have a unique constraint on identifier, I think this is pretty safe given the ORM docs and how we test `upsert` throughout the rest of the API. If you feel strongly though, I can update the test.

Reply:
All good, I just wasn't sure, thanks for context